collagraph/renderers/pyside_renderer.py

```diff
@@ -0,0 +1,181 @@
+from collections import defaultdict
+import logging
+from typing import Any, Callable
+
+from PySide6 import QtCore, QtWidgets
+from PySide6.QtWidgets import (
+    QBoxLayout,
+    QDialogButtonBox,
+    QFormLayout,
+    QGridLayout,
+    QMainWindow,
+    QSpacerItem,
+    QTabWidget,
+    QWidget,
+)
+
+from . import Renderer
+from .pyside import camel_case, dialog, name_to_type, tab, widget, window
+
+
+logger = logging.getLogger(__name__)
+
+
+def sorted_on_class_hierarchy(value):
+    # __mro__ is a tuple of all the classes that an class inherits. The longer the
+    # __mro__, the 'deeper' the subclass is, so we can use that to sort the classes
+    # to make the deepest class come first.
+    return dict(sorted(value.items(), key=lambda x: -len(x[0].__mro__)))
+
+
+INSERT_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.insert,
+        QMainWindow: window.insert,
+        QDialogButtonBox: dialog.insert,
+        QTabWidget: tab.insert,
+    }
+)
+REMOVE_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.remove,
+        QTabWidget: tab.remove,
+    }
+)
+SET_ATTR_MAPPING = sorted_on_class_hierarchy(
+    {
+        QWidget: widget.set_attribute,
+    }
+)
+
+# Cache for wrapped types
+WRAPPED_TYPES = {}
+
+LAYOUT = {
+    "Box": QBoxLayout,
+    "Grid": QGridLayout,
+    "Form": QFormLayout,
+}
+
+
+def not_implemented(self, *args, **kwargs):
+    raise NotImplementedError(type(self).__name__)
+
+
+class PySideRenderer(Renderer):
+    """PySide6 renderer."""
+
+    def create_element(self, type_name: str) -> Any:
+        """Create an element for the given type."""
+        # Make sure that an app exists before any widgets
+        # are created. Otherwise we might experience a
+        # hard segfault.
+        QtCore.QCoreApplication.instance() or QtWidgets.QApplication()
```

Comment:
just thinking you might want to cache this check on `self._app` or something, because I'm thinking this is a pretty slow operation in comparison to the other stuff that's happening in this function and the amount of calls is going to add up at some point